lib/Frontend/CachedDiagnostics.cpp

```diff
@@ -617,7 +617,7 @@ llvm::Error DiagnosticSerializer::deserializeDiagnosticInfo(
                                   Kind,
                                   Info.FormatString,
                                   {},
-                                  Info.Category,
+                                  Info.Category.empty() ? StringRef() : Info.Category,
```

Comment:
I can't see what difference it makes? Can you explain?

Reply:
The parameter expects an optional pointer when passed to bridging so without this we'd happy pass a valid empty string instead of `nil` and that leads to swift-syntax printing `[#]` as a category because the expectation is that a valid category always has a name.

Reply:
That was added here - https://github.com/swiftlang/swift-syntax/pull/2981/files#diff-0a73410f2b8e6bd296483c34ce6f07228312dffe3eaafc520ef4bae5f0cbe157R54

Reply:
Is this because the `StringRef` constructed from an empty string doesn't have `data()` to be `nullptr`?

This is a very unexpected result and very hard to understand here. Is it possible to get `swift_ASTGen_addQueuedDiagnostic` to bridge the empty StringRef as `nil`?

Reply:
Or `DiagnosticBridge.cpp` will pass nullptr instead? Doing it here is a bit hard to understand

Reply:
I'm not sure what you mean, we are using a `SerializedDiagnosticInfo` that has std::string for this field, empty std::string is not a nullptr, that's why we are in this situation.

Reply:
Right, that is what I understand, but I wonder if it should be bridged elsewhere instead.

If category can ever be empty, then the bridging in `swift_ASTGen_addQueuedDiagnostic` should just treat as zero length stringref as nil.
If category can be empty, then the abstraction in DiagnosticInfo needs to be updated, together with this serialization format.

Reply:
I don't think that is always the case, it's definitely is for category though, in generally I doubt that.